### PR TITLE
SI-7265 Buy some time to fix s.u.Properties.isJavaAtLeast

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

File: src/library/scala/util/Properties.scala [hunk 1]
```diff
@@ -150,16 +150,32 @@ private[scala] trait PropertiesTrait {
...
+    *
+    * @example {{{
+    * // In this example, the runtime's Java version is considered to be 1.7.
+    * isJavaAtLeast("1.6")            // true
+    * isJavaAtLeast("1.7")            // true
+    * isJavaAtLeast("1.8")            // false
+    * }}
+    */
   def isJavaAtLeast(version: String) = {
+    val     v9 = "1.9" ::   Nil
```
Comment (adriaanm) [inline src/library/scala/util/Properties.scala:167]: Isn't this all a bit roundabout?

```
def isJavaAtLeast(version: String) = version.length > 2 && version.startsWith("1.") && javaVersion(2) >= version(2)
```

Comment (adriaanm) [inline src/library/scala/util/Properties.scala:167]: yes, one could imagine doing a bit better, especially since 1.10 is around the corner...

Comment (som-snytt) [inline src/library/scala/util/Properties.scala:167]: It is seeking a middle ground with the previous attempt that compared the entire version string using bit masking; the details escape me.  I've forgotten why split was deemed too high-level?

On simplicity: I wouldn't mind an API that asked 

`isJavaAtLeast(major: Int, minor: Int = 0, epoch: Int = 1)` and `isJavaAtLeast(5)`.

That jibes with the Scala ethos of not buying into every historical fiction in Java.

And no one will complain that their "1.7.0_36b3" lied to them.

+1 for leaving our grandchildren out of it, who may very well prefer the term transcircuitous to cyborg; and also omitting any reference to Gosling's Law, that the rate of progress of major versions is inversely proportional to the square of the sum of LOC + features + open bugs.

Comment (soc) [inline src/library/scala/util/Properties.scala:167]: > that compared the entire version string using bit masking 

Wow ... I just used some while loops! :-)

**Comment by paulp** [comment]
"I've forgotten why split was deemed too high-level?"

split is the same as using a regex, and apparently regexes drag in a lot.

**Comment by adriaanm** [comment]
Going to close this PR this week unless it gets updated. It's had pending review comments for over 10 days.

**Comment by soc** [comment]
@adriaanm Sorry, broken nose, hospital, etc... Reading through the comments, I don't see stuff which would be applicable to the issue here. (I sympathize with the version which takes integers instead of a string, but that doesn't really fix the issues with the string-based one, does it?)

**Comment by adriaanm** [comment]
@soc, sorry to hear it! Hope it's healing well!

**Comment by adriaanm** [comment]
I'm still in favor of my oneliner as opposed to the convoluted status quo (before this PR).

**Comment by som-snytt** [comment]
@soc my condolences, too.  I guess you have to be pretty careful how you give feedback on github, you never know how it will be received.  I'd also suggest preferring one-line to involution or convolution. The one line does give different answers for `"1.*"` and `"1.?"`.  Maybe the `1.*` is a typo for `1.8`?

**Comment by soc** [comment]
The thing is that if you want to make that one-liner work reliably, you end up with exactly the change which was rejected before.

So we have either 
- Something easy which will break
- Something imperative, which works, but was already rejected
- Something which uses regexes and pulls in tons of dependencies

**Comment by som-snytt** [comment]
Here is something between a 1-liner, the 3-liner requested by @paulp and the original forty-niner I mean forty-liner.

https://gist.github.com/som-snytt/5725336

There's a shorter take that assumes major.minor, and a longer version that takes as many parts as you want to try to compare against whatever the actual version string is.  I preserve the original strategy of taking the numeric prefix of the part for purposes of comparison, so "0_52" is 0, "5b" is 5, "b" is 0.

I think somewhere in the repo is version munging logic for the sbt build, or similar endeavor.

**Comment by soc** [comment]
I don't want to imagine how many classes this creates for something as trivial as comparing a version number ...

**Comment by paulp** [comment]
Hey everyone, what are we doing?

```
scala> System.getProperty("java.specification.version")
res0: String = 1.6
```

**Comment by som-snytt** [comment]
```
def isScalaAtLeast(spec: String) = spec == "2.9"
```

A little spec humo(u)r.

**Comment by paulp** [comment]
So @soc would you like to take a shot at this based on java.specification.version, which makes it easy unless you're trying to future-proof against java 10.x.

```
scala> sys.props("java.specification.version") match { case s => (s.substring(0, 1).toInt, s.substring(2).toInt) }
res0: (Int, Int) = (1,6)
```

**Comment by gkossakowski** [comment]
@soc: what's the status of this PR?

**Comment by som-snytt** [comment]
@gkossakowski Without @paulp 's help we can't figure out how to perform the calculation in acceptable time and space such that the code is maintainably write-once read-many!  But maybe @soc will bounce back.

Maybe there's a scala.reflect.internal.util.ReflectionUtils.isJavaAtLeast method that could be made public.

**Comment by soc** [comment]
Not sure about a good solution ... I'll close it and will reopen if I or someone else has a new idea.

**Comment by soc** [comment]
Wait, isn't this just the temporary fix?
Ok, my suggestion is to merge it as-is (it is just adding a few version strings), until we figure out a nicer implementation.

**Comment by som-snytt** [comment]
I'll have a PR in a minute after I figure out how to submit against 2.10 under the new github interface, argh.  The PR includes a junit test! ha!  That's great.  And corrects the scaladoc in this PR, which is backward. :8ball:  I'm not sure I'm using 8ball correctly, but github suggested it.  I kind of want my old github back.

**Pull request closed without merging**